src/mlpack/methods/kmeans/naive_kmeans_impl.hpp

```diff
@@ -29,75 +33,115 @@ NaiveKMeans<DistanceType, MatType>::NaiveKMeans(const MatType& dataset,
+
+  // Pre-allocate thread-local storage
+  std::vector<arma::mat> threadCentroids(effectiveThreads, arma::mat(dims, clusters, arma::fill::zeros));
+  std::vector<arma::Col<size_t>> threadCounts(effectiveThreads, arma::Col<size_t>(clusters, arma::fill::zeros));
 
-    #pragma omp for
-    for (size_t i = 0; i < (size_t) dataset.n_cols; ++i)
+  #ifdef MLPACK_USE_OPENMP
+    #pragma omp parallel num_threads(effectiveThreads)
+  #endif
+  {
+    size_t threadId = 0;
+    #ifdef MLPACK_USE_OPENMP
+      threadId = omp_get_thread_num();
+    #endif
+    const size_t segmentStart = threadId * nominalSegmentSize;
+    const size_t segmentEnd = (threadId == effectiveThreads - 1) ? points : (threadId + 1) * nominalSegmentSize;
+
+    arma::mat& localCentroids = threadCentroids[threadId];
+    arma::Col<size_t>& localCounts = threadCounts[threadId];
+
+    arma::vec distances(clusters);
+
+    for (size_t i = segmentStart; i < segmentEnd; ++i)
     {
-      // Find the closest centroid to this point.
-      double minDistance = std::numeric_limits<double>::infinity();
-      size_t closestCluster = centroids.n_cols; // Invalid value.
+      const auto dataPoint = dataset.col(i);
+      const double dataNorm = arma::dot(dataPoint, dataPoint);
 
-      for (size_t j = 0; j < centroids.n_cols; ++j)
+      // Calculate distances to all centroids
+      for (size_t j = 0; j < clusters; ++j)
       {
-        const double dist = distance.Evaluate(dataset.col(i),
-            centroids.unsafe_col(j));
-        if (dist < minDistance)
-        {
-          minDistance = dist;
-          closestCluster = j;
-        }
+        const arma::vec& centroid = centroids.col(j);
+        distances(j) = std::max(0.0, dataNorm + centroidNorms(j) - 2 * arma::dot(dataPoint, centroid));
       }
 
-      Log::Assert(closestCluster != centroids.n_cols);
+      // Find the closest centroid
+      const size_t closestCluster = distances.index_min();
 
-      // We now have the minimum distance centroid index.  Update that centroid.
-      localCentroids.unsafe_col(closestCluster) += dataset.col(i);
+      // Update local centroids and counts
+      localCentroids.col(closestCluster) += dataPoint;
       localCounts(closestCluster)++;
     }
-    // Combine calculated state from each thread
-    #pragma omp critical
-    {
-      newCentroids += localCentroids;
-      counts += localCounts;
-    }
   }
 
-  // Now normalize the centroid.
-  for (size_t i = 0; i < centroids.n_cols; ++i)
-    if (counts(i) != 0)
-      newCentroids.col(i) /= counts(i);
+  // Combine results from all threads
+  for (size_t t = 0; t < effectiveThreads; ++t)
+  {
+    newCentroids += threadCentroids[t];
+    counts += threadCounts[t];
+  }
 
-  distanceCalculations += centroids.n_cols * dataset.n_cols;
+  const double eps = std::numeric_limits<double>::epsilon();
+  // Normalize the centroids
+  for (size_t j = 0; j < clusters; ++j)
+  {
+    if (counts(j) > eps)
```

Comment:
But `counts(j)` is a `size_t` and so you would want to compare it with `0`.  (The previous version of this loop was just fine, as far as I can tell.)